tests/modules/apigee_mock_client/src/MockHandlerStack.php

```diff
@@ -0,0 +1,164 @@
+   *
+   * @var array
+   */
+  protected $responses;
+
+  /**
+   * The twig environment for getting response data.
+   *
+   * @var \Twig_Environment
+   */
+  protected $twig;
+
+  /**
+   * @var \Drupal\Core\Queue\QueueInterface
+   */
+  protected $database_queue;
+
+  /**
+   * @param \Twig_Environment $twig
+   * @param \Drupal\Core\Queue\QueueFactory $queue_factory
+   */
+  public function __construct(\Twig_Environment $twig, QueueFactory $queue_factory) {
+    $this->twig = $twig;
+    $this->database_queue = $queue_factory->get('apigee_edge_mock_responses', TRUE);
+
+    parent::__construct();
+  }
+
+  /**
+   * Queue a response that is in the catalog. Dynamic values can be passed and
+   * will be replaced in the response. For example `['foo' => [':bar' => 'baz']]`
+   * will load the catalog entry named `foo `and replace `:bar` with "baz" in the
+   * body text.
+   *
+   * @param string|array $response_ids
+   *   The response id to queue in one of the following formats:
+   *     - `'foo'`
+   *     - `['foo', 'bar']` // Queue multiple responses.
+   *     - `['foo' => [':bar' => 'baz']]` // w/ replacements for response body.
+   *
+   * @return $this
+   *
+   * @throws \Twig_Error_Loader
+   * @throws \Twig_Error_Runtime
+   * @throws \Twig_Error_Syntax
+   */
+  public function queueFromResponseFile($response_ids) {
+    $org_name = \Drupal::service('apigee_edge.sdk_connector')->getOrganization();
+
+    if (empty($this->responses)) {
+      // Get the module path for this module.
+      $module_path = \Drupal::moduleHandler()->getModule('apigee_mock_client')->getPath();
+      $this->responses = Yaml::parseFile($module_path . '/response_catalog.yml')['responses'];
+    }
+    // Loop through responses and add each one.
+    foreach ((array) $response_ids as $index => $item) {
+      // The catalog id should either be the item itself or the keys if an associative array has been passed.
+      $id = !is_array($item) ? $item : $index;
+      // Body text can have elements replaced in it for certain values. See: http://php.net/strtr
+      $context = is_array($item) ? $item : [];
+      $context['org_name'] = isset($context['org_name']) ? $context['org_name'] : $org_name;
+
+      // Add the default headers if headers aren't defined in the response catalog.
+      $headers = isset($this->responses[$id]['headers']) ? $this->responses[$id]['headers'] : [
+        'content-type' => 'application/json;charset=utf-8'
+      ];
+      // Set the default status code.
+      $status_code = !empty($this->responses[$id]['status_code']) ? $this->responses[$id]['status_code'] : 200;
+      $status_code = !empty($context['status_code']) ? $context['status_code'] : $status_code;
+
+      // Render the response content.
+      $template = str_replace('_', '-', $id) . '.json.twig';
+      $content = $this->twig->getLoader()->exists($template) ? $this->twig->render($template, $context) : '';
```

Comment:
How other modules can alter the response ids, status code and/or template file before the response object is being constructed?

Reply:
There shouldn't be a need to alter response ids or the template file. The status code can be passed in i.e. `queueMockResponse(['foo' => ['status_code' => 201]])`.